File: Bulwark.Service/Reputation/ThreatFoxFeed.cs
using System.Text;
using System.Text.Json;
using Bulwark.Core.Models;
using Microsoft.Extensions.Logging;

namespace Bulwark.Service.Reputation;

/// <summary>ThreatFox 返回的一条 IOC 记录(仅取生成规则需要的字段)。</summary>
public sealed record ThreatFoxIoc(
    string Ioc,          // IOC 值(如 "1.2.3.4:443"、"evil.com"、sha256)
    string IocType,      // ip:port / domain / url / md5_hash / sha256_hash
    string? Malware,     // 恶意家族可读名
    string? ThreatType,  // 威胁类型(如 botnet_cc、payload)
    int Confidence);     // 置信度 0-100

/// <summary>
/// ThreatFox(abuse.ch)情报 feed 客户端 + 规则生成器。
///
/// 定期批量拉取最近已知恶意 IOC(get_iocs),转换成一批 <see cref="DefenseRule"/>:
///  · sha256_hash → 按哈希 Block(整文件禁跑,改名无效);
///  · ip:port     → NetworkConnect Block(拦截外联到该 IP);
///  · domain      → NetworkConnect Block(可选)。
///
/// 网络经系统 curl.exe 发起(与其它信誉源一致,规避本环境 .NET TLS 握手问题)。
/// 任何失败都返回空列表/不抛断,绝不影响主防护流程。
/// </summary>
public sealed class ThreatFoxFeedClient
{
    private const string ApiUrl = "https://threatfox-api.abuse.ch/api/v1/";

    private readonly ILogger _logger;
    private readonly ThreatFoxFeedOptions _opt;
    private readonly string _authKey;

    public bool IsEnabled => _opt.Enabled && !string.IsNullOrWhiteSpace(_authKey);

    public ThreatFoxFeedClient(ILogger logger, ThreatFoxFeedOptions opt, string? malwareBazaarKey)
    {
        _logger = logger;
        _opt = opt;
        _authKey = opt.ResolveAuthKey(malwareBazaarKey);
    }

    /// <summary>拉取最近 <see cref="ThreatFoxFeedOptions.Days"/> 天、置信度达标的 IOC。失败返回空列表。</summary>
    public async Task<IReadOnlyList<ThreatFoxIoc>> FetchRecentAsync(CancellationToken token = default)
    {
        if (!IsEnabled) return Array.Empty<ThreatFoxIoc>();

        int days = Math.Clamp(_opt.Days, 1, 7);
        string body = JsonSerializer.Serialize(new { query = "get_iocs", days });

        try
        {
            var (code, resp) = await RunCurlAsync(body, token);
            if (code != 200)
            {
                _logger.LogWarning("ThreatFox feed 拉取失败,HTTP {code}。", code);
                return Array.Empty<ThreatFoxIoc>();
            }
            return Parse(resp);
        }
        catch (OperationCanceledException) { return Array.Empty<ThreatFoxIoc>(); }
        catch (Exception ex)
        {
            _logger.LogDebug(ex, "ThreatFox feed 拉取异常。");
            return Array.Empty<ThreatFoxIoc>();
        }
    }

    private List<ThreatFoxIoc> Parse(string json)
    {
        var list = new List<ThreatFoxIoc>();
        try
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (!root.TryGetProperty("query_status", out var qs)
                || !string.Equals(qs.GetString(), "ok", StringComparison.OrdinalIgnoreCase))
                return list;
            if (!root.TryGetProperty("data", out var data) || data.ValueKind != JsonValueKind.Array)
                return list;

            foreach (var it in data.EnumerateArray())
            {
                string ioc = GetStr(it, "ioc");
                string iocType = GetStr(it, "ioc_type");
                if (string.IsNullOrWhiteSpace(ioc) || string.IsNullOrWhiteSpace(iocType)) continue;

                int conf = it.TryGetProperty("confidence_level", out var c) && c.TryGetInt32(out var n) ? n : 0;
                if (conf < _opt.MinConfidence) continue;

                string? malware = GetStrOrNull(it, "malware_printable") ?? GetStrOrNull(it, "malware");
                string? threat = GetStrOrNull(it, "threat_type");
                list.Add(new ThreatFoxIoc(ioc.Trim(), iocType.Trim().ToLowerInvariant(), malware, threat, conf));
            }
        }
        catch (Exception ex)
        {
            _logger.LogDebug(ex, "解析 ThreatFox feed 响应失败。");
        }
        return list;
    }

    private static string GetStr(JsonElement e, string name)
        => e.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() ?? "" : "";

    private static string? GetStrOrNull(JsonElement e, string name)
        => e.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    /// <summary>经 curl.exe POST(带 Auth-Key + application/json),返回 (状态码, 响应体)。</summary>
    private async Task<(int Code, string Body)> RunCurlAsync(string jsonBody, CancellationToken token)
    {
        var psi = new System.Diagnostics.ProcessStartInfo
        {
            FileName = "curl.exe",
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true,
            StandardOutputEncoding = Encoding.UTF8,
        };
        psi.ArgumentList.Add("-sS");
        psi.ArgumentList.Add("-k");
        psi.ArgumentList.Add("--max-time");
        psi.ArgumentList.Add(Math.Max(10, _opt.QueryTimeoutSeconds).ToString());
        psi.ArgumentList.Add("-H");
        psi.ArgumentList.Add("Auth-Key: " + _authKey);
        psi.ArgumentList.Add("-H");
        psi.ArgumentList.Add("Content-Type: application/json");
        psi.ArgumentList.Add("--data-raw");
        psi.ArgumentList.Add(jsonBody);
        psi.ArgumentList.Add("-w");
        psi.ArgumentList.Add("\nHTTPSTATUS:%{http_code}");
        psi.ArgumentList.Add(ApiUrl);

        using var proc = new System.Diagnostics.Process { StartInfo = psi };
        if (!proc.Start()) return (0, string.Empty);

        var outTask = proc.StandardOutput.ReadToEndAsync(token);
        var errTask = proc.StandardError.ReadToEndAsync(token);
        await proc.WaitForExitAsync(token);
        var stdout = await outTask;
        _ = await errTask;

        int code = 0;
        string bodyText = stdout;
        const string marker = "\nHTTPSTATUS:";
        int idx = stdout.LastIndexOf(marker, StringComparison.Ordinal);
        if (idx >= 0)
        {
            bodyText = stdout[..idx];
            int.TryParse(stdout[(idx + marker.Length)..].Trim(), out code);
        }
        return (code, bodyText);
    }
}

/// <summary>
/// 把 ThreatFox 拉到的 IOC 批量转换成防护规则(<see cref="DefenseRule"/>)。
/// 纯函数、无副作用;去重、上限、过期时间、来源标记都在此处理。
/// </summary>
public static class IntelRuleGenerator
{
    /// <summary>
    /// 生成规则列表。规则统一带来源标记(便于刷新时先清旧再灌新)与过期时间。
    /// </summary>
    public static List<DefenseRule> Generate(IReadOnlyList<ThreatFoxIoc> iocs, ThreatFoxFeedOptions opt)
    {
        var rules = new List<DefenseRule>();
        if (iocs is null || iocs.Count == 0) return rules;

        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        DateTime expires = DateTime.UtcNow.AddDays(Math.Max(1, opt.RuleTtlDays));
        int max = Math.Max(1, opt.MaxRules);

        foreach (var ioc in iocs)
        {
            if (rules.Count >= max) break;
            if (!seen.Add(ioc.IocType + "|" + ioc.Ioc)) continue;

            var rule = ioc.IocType switch
            {
                "sha256_hash" when opt.GenerateHashRules => BuildHashRule(ioc, expires),
                "ip:port" when opt.GenerateIpRules       => BuildIpRule(ioc, expires),
                "domain" when opt.GenerateDomainRules    => BuildDomainRule(ioc, expires),
                _ => null,
            };
            if (rule is not null) rules.Add(rule);
        }
        return rules;
    }

    private static string Label(ThreatFoxIoc ioc)
    {
        string fam = !string.IsNullOrWhiteSpace(ioc.Malware) ? ioc.Malware!
                   : !string.IsNullOrWhiteSpace(ioc.ThreatType) ? ioc.ThreatType!
                   : "malicious";
        return $"{ThreatFoxFeedOptions.RuleNoteTag} {fam} ({ioc.Confidence}%)";
    }

    /// <summary>SHA-256 → 按哈希 Block:Type=null 表示该文件的任何行为都拦截,改名无效。</summary>
    private static DefenseRule BuildHashRule(ThreatFoxIoc ioc, DateTime expires) => new()
    {
        ActorHashes = new HashSet<string>(StringComparer.OrdinalIgnoreCase) { ioc.Ioc.ToUpperInvariant() },
        Type = null,
        Action = VerdictAction.Block,
        HardOverride = true,
        ExpiresUtc = expires,
        Note = Label(ioc),
    };

    /// <summary>ip:port → NetworkConnect Block:拦截外联到该 IP(任意端口)。</summary>
    private static DefenseRule BuildIpRule(ThreatFoxIoc ioc, DateTime expires)
    {
        string ip = ioc.Ioc;
        int colon = ip.LastIndexOf(':');
        if (colon > 0) ip = ip[..colon];   // 去掉 :port,拦该 IP 的任意端口
        return new DefenseRule
        {
            Type = EventType.NetworkConnect,
            TargetPattern = ip.Trim() + "*",
            Action = VerdictAction.Block,
            ExpiresUtc = expires,
            Note = Label(ioc),
        };
    }

    /// <summary>domain → NetworkConnect Block:目标中包含该域名即拦。</summary>
    private static DefenseRule BuildDomainRule(ThreatFoxIoc ioc, DateTime expires) => new()
    {
        Type = EventType.NetworkConnect,
        TargetPattern = "*" + ioc.Ioc.Trim() + "*",
        Action = VerdictAction.Block,
        ExpiresUtc = expires,
        Note = Label(ioc),
    };
}

File: cpp/service/src/reputation/ThreatFoxFeed.cpp
#include "bulwark/service/reputation/ThreatFoxFeed.h"
#include "bulwark/service/reputation/ReputationCurl.h"
// 「这个 IP 能不能整段封禁」的统一判定,与 Worker::buildRulesFromProfile 共用同一份名单。
#include "bulwark/service/IpBlockPolicy.h"

#include <QCoreApplication>
#include <QDateTime>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonValue>
#include <QSet>

#include <algorithm>
#include <chrono>

namespace bulwark::service::reputation {
namespace {
inline QString u(const char* s) { return QString::fromUtf8(s); }
} // namespace

// ============================ ThreatFoxFeedClient ============================

ThreatFoxFeedClient::ThreatFoxFeedClient(const ThreatFoxFeedOptions& opt, const QString& malwareBazaarKey)
    : opt_(opt), authKey_(opt.resolveAuthKey(malwareBazaarKey)) {}

bool ThreatFoxFeedClient::isEnabled() const {
    return opt_.Enabled && !authKey_.trimmed().isEmpty();
}

QVector<ThreatFoxIoc> ThreatFoxFeedClient::fetchRecent() {
    if (!isEnabled())
        return {};

    const int days = std::clamp(opt_.Days, 1, 7);
    QJsonObject req;
    req[QStringLiteral("query")] = QStringLiteral("get_iocs");
    req[QStringLiteral("days")] = days;
    const QString body = QString::fromUtf8(QJsonDocument(req).toJson(QJsonDocument::Compact));

    const QStringList headers{
        QStringLiteral("Auth-Key: ") + authKey_,
        QStringLiteral("Content-Type: application/json") };
    const QString url = opt_.BaseUrl.isEmpty() ? QStringLiteral("https://threatfox-api.abuse.ch/api/v1/")
                                               : opt_.BaseUrl;
    const auto res = ReputationCurl::postRaw(url, body, headers, std::max(10, opt_.QueryTimeoutSeconds));
    if (res.first != 200)
        return {};
    return parse(res.second);
}

QVector<ThreatFoxIoc> ThreatFoxFeedClient::parse(const QString& json) const {
    QVector<ThreatFoxIoc> list;
    QJsonParseError err{};
    const QJsonDocument doc = QJsonDocument::fromJson(json.toUtf8(), &err);
    if (err.error != QJsonParseError::NoError || !doc.isObject())
        return list;
    const QJsonObject root = doc.object();
    if (root.value(QLatin1String("query_status")).toString().compare(QLatin1String("ok"), Qt::CaseInsensitive) != 0)
        return list;
    const QJsonArray data = root.value(QLatin1String("data")).toArray();
    for (const QJsonValue& v : data) {
        const QJsonObject it = v.toObject();
        const QString ioc = it.value(QLatin1String("ioc")).toString().trimmed();
        const QString iocType = it.value(QLatin1String("ioc_type")).toString().trimmed().toLower();
        if (ioc.isEmpty() || iocType.isEmpty())
            continue;
        const QJsonValue confVal = it.value(QLatin1String("confidence_level"));
        const int conf = confVal.isDouble() ? confVal.toInt() : 0;
        if (conf < opt_.MinConfidence)
            continue;
        ThreatFoxIoc rec;
        rec.ioc = ioc;
        rec.iocType = iocType;
        rec.malware = it.value(QLatin1String("malware_printable")).toString();
        if (rec.malware.isEmpty())
            rec.malware = it.value(QLatin1String("malware")).toString();
        rec.threatType = it.value(QLatin1String("threat_type")).toString();
        rec.confidence = conf;
        list.append(rec);
    }
    return list;
}

// ============================ IntelRuleGenerator =============================

QVector<bulwark::DefenseRule> generateIntelRules(const QVector<ThreatFoxIoc>& iocs,
                                                 const ThreatFoxFeedOptions& opt) {
    QVector<bulwark::DefenseRule> rules;
    if (iocs.isEmpty())
        return rules;

    QSet<QString> seen;
    const QDateTime expires = QDateTime::currentDateTimeUtc().addDays(std::max(1, opt.RuleTtlDays));
    const int maxRules = std::max(1, opt.MaxRules);
    const QString tag = ThreatFoxFeedOptions::ruleNoteTag();

    const auto label = [&tag](const ThreatFoxIoc& i) -> QString {
        const QString fam = !i.malware.trimmed().isEmpty()   ? i.malware
                          : !i.threatType.trimmed().isEmpty() ? i.threatType
                                                              : QStringLiteral("malicious");
        return tag + QLatin1Char(' ') + fam + QStringLiteral(" (") + QString::number(i.confidence)
             + QStringLiteral("%)");
    };

    for (const ThreatFoxIoc& i : iocs) {
        if (rules.size() >= maxRules)
            break;
        const QString key = i.iocType + QLatin1Char('|') + i.ioc;
        if (seen.contains(key))
            continue;
        seen.insert(key);

        bulwark::DefenseRule rule;
        bool ok = false;
        if (i.iocType == QLatin1String("sha256_hash") && opt.GenerateHashRules) {
            // 按哈希 Block:type 留空(任意行为都拦),改名无效;硬覆盖优先级最高。
            rule.actorHashes.insert(i.ioc.toUpper());
            rule.action = bulwark::VerdictAction::Block;
            rule.hardOverride = true;
            ok = true;
        } else if (i.iocType == QLatin1String("ip:port") && opt.GenerateIpRules) {
            QString ip = i.ioc;
            const int colon = ip.lastIndexOf(QLatin1Char(':'));
            if (colon > 0)
                ip = ip.left(colon); // 去 :port,拦该 IP 任意端口
            ip = ip.trimmed();
            //
            // 共享基础设施不做整段封禁(与 Worker::buildRulesFromProfile 同一判定)。
            // feed 给出的置信度是「这个 IOC 与该家族相关」的置信度,不是「整段封掉这个 IP
            // 是安全的」的置信度 —— C2 面板套在 Cloudflare / Edgecast 后面很常见,按 IP 封会
            // 把该前端上成千上万个无关站点一起打死。实测本机就有 4 条 192.229.115.x
            //(Edgecast)的 PureRAT 规则,属于这种情况。
            //
            if (isUnsafeToBlanketBlockIp(ip))
                continue;
            rule.type = bulwark::EventType::NetworkConnect;
            //
            // 【必须带冒号】。原来是 ip + '*',而 '*' 紧跟在末段数字之后 —— "154.12.94.16*"
            // 除了 154.12.94.16 还会一并匹配 154.12.94.160 ~ 154.12.94.169,凭空多封 10 个
            // 无关地址。加上 ':' 把通配锚定在「端口」位置,语义才是「该 IP 的任意端口」,
            // 也与行为画像分支产出的 "<ip>:*" 保持一致。
            //
            rule.targetPattern = ip + QStringLiteral(":*");
            rule.action = bulwark::VerdictAction::Block;
            ok = true;
        } else if (i.iocType == QLatin1String("domain") && opt.GenerateDomainRules) {
            rule.type = bulwark::EventType::NetworkConnect;
            rule.targetPattern = QLatin1Char('*') + i.ioc.trimmed() + QLatin1Char('*');
            rule.action = bulwark::VerdictAction::Block;
            ok = true;
        }
        if (ok) {
            rule.expiresUtc = expires;
            rule.note = label(i);
            rules.append(rule);
        }
    }
    return rules;
}

// ============================ IntelFeedService ===============================

IntelFeedService::IntelFeedService(const ThreatFoxFeedOptions& opt, const QString& malwareBazaarKey)
    : client_(opt, malwareBazaarKey), opt_(opt) {}

IntelFeedService::~IntelFeedService() { stop(); }

bool IntelFeedService::isEnabled() const { return client_.isEnabled(); }

bool IntelFeedService::sleepInterruptible(int seconds) {
    if (seconds <= 0)
        return running_.load();
    std::unique_lock<std::mutex> lk(mx_);
    cv_.wait_for(lk, std::chrono::seconds(seconds), [this] { return !running_.load(); });
    return running_.load();
}

void IntelFeedService::loop() {
    // 首次延迟(避开开机拥塞),期间被 stop 唤醒则直接退出。
    if (!sleepInterruptible(std::max(0, opt_.InitialDelaySeconds)))
        return;

    while (running_.load()) {
        const QVector<ThreatFoxIoc> iocs = client_.fetchRecent();
        if (!iocs.isEmpty() && onRulesReady_) {
            const QVector<bulwark::DefenseRule> rules = generateIntelRules(iocs, opt_);
            if (!rules.isEmpty()) {
                log_.info(u("ThreatFox 情报:拉取 ") + QString::number(iocs.size())
                          + u(" 条 IOC,生成 ") + QString::number(rules.size()) + u(" 条规则,编组注入。"));
                RulesReadyCallback cb = onRulesReady_;
                // 编组回主线程再注入引擎/落盘(避免跨线程碰 RuleStore / 保持与其它引擎变更一致)。
                QMetaObject::invokeMethod(
                    qApp, [cb, rules] { cb(rules); }, Qt::QueuedConnection);
            }
        }
        if (opt_.RefreshIntervalHours <= 0)
            break; // 仅启动时拉一次
        if (!sleepInterruptible(opt_.RefreshIntervalHours * 3600))
            return;
    }
}

void IntelFeedService::start() {
    if (!client_.isEnabled()) {
        log_.info(u("ThreatFox 情报 feed 未启用(未开启或无 Auth-Key),后台服务不启动。"));
        return;
    }
    if (running_.exchange(true))
        return;
    worker_ = std::thread([this] { loop(); });
    log_.info(u("ThreatFox 情报 feed 后台服务已启动。"));
}

void IntelFeedService::stop() {
    if (!running_.exchange(false))
        return;
    {
        std::lock_guard<std::mutex> lk(mx_);
        cv_.notify_all();
    }
    if (worker_.joinable())
        worker_.join();
}

} // namespace bulwark::service::reputation
